windows/devices/uwb/include/windows/devices/uwb/UwbConnector.hxx

```diff
@@ -226,15 +221,52 @@ private:
     std::vector<std::shared_ptr<::uwb::UwbRegisteredDeviceEventCallbacks>>
     GetResolvedDeviceEventCallbacks();
 
+    /**
+     * @brief Internal helper function to insert the callback into the hashmap of ids to callbacks
+     * You MUST have grabbed the m_eventCallbacksGate mutex prior to calling this.
+     *
+     * @param callback
+     * @return std::weak_ptr<::uwb::RegisteredCallbackToken>
+     */
+    std::shared_ptr<::uwb::RegisteredCallbackToken>
+    InsertDeviceEventCallback(std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks> callback);
+
+    /**
+     * @brief Internal helper function to insert the callback into the hashmap of ids to callbacks
+     * You MUST have grabbed the m_eventCallbacksGate mutex prior to calling this.
+     *
+     * @param callback
+     * @return std::weak_ptr<::uwb::RegisteredCallbackToken>
+     */
+    std::shared_ptr<::uwb::RegisteredCallbackToken>
+    InsertSessionEventCallback(std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks> callback);
+
+    /**
+     * @brief Internal helper function that deregisters event callbacks.
+     * If you pass in a token that is no longer valid, this function does nothing
+     * You must have grabbed the m_eventCallbacksGate mutex prior to calling this
+     *
+     * @param token
+     */
+    void
+    DeregisterEventCallback(::uwb::RegisteredCallbackToken token);
+
 private:
     // the following shared_mutex is used to protect access to both session and device event callbacks
     mutable std::shared_mutex m_eventCallbacksGate;
-    std::unordered_map<uint32_t, std::vector<std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks>>> m_sessionEventCallbacks;
-    std::vector<std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks>> m_deviceEventCallbacks;
+    std::unordered_map<uint32_t, std::vector<uint32_t>> m_sessionEventCallbacks;
+    std::vector<uint32_t> m_deviceEventCallbacks;
     std::string m_deviceName{};
     std::jthread m_notificationThread;
     wil::shared_hfile m_notificationHandleDriver;
     OVERLAPPED m_notificationOverlapped;
+
+private:
+    // the following is related to how the UwbConnector keeps track of its callbacks
+    uint32_t m_tokenUniqueState;
+    std::unordered_map<uint32_t, std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks>> m_sessionEventCallbacksIdMap;
+    std::unordered_map<uint32_t, std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks>> m_deviceEventCallbacksIdMap;
+    std::vector<std::shared_ptr<::uwb::RegisteredCallbackToken>> m_tokens;

```

Comment:
There's a lot of state here and having multiple maps makes it complicated to track. A few observations:

* `m_uniqueTokenState` is used to uniquely identify a specific callback registration. For each such registration, we're creating a `std::shared_ptr<RegisteredCallbackToken>`. So, we could do at least one of two things here to reduce the complexity and consolidate: 
    - Move the uniqueness into `RegisteredCallbackToken` directly, where it has a `private static` function to generate a unique identifier and assigns it to `callbackId` upon construction. This could just use a `static std::atomic<uint32_>` to generate new unique identifiers.
    - Remove `callbackId` all-together and instead use the fact that the memory address of the `shared_ptr<RegisteredCallbackToken>` (ie. that obtained from `.get()`) is itself unique and could be used to look up the token. Eg.

```
void UwbConnector::DeregisterEventCallback(std::weak_ptr<::uwb::RegisteredCallbackToken> tokenWeak) {
    auto tokenToDeregister = tokenWeak.lock():
    auto tokenIt = std::find_if(m_tokens, [](const auto& token) { return token.get() == tokenToDeregister.get(); });
    if (tokenIt == std::cend(m_tokens)) {
        return; // no associated token found, bail
    }
    std::shared_ptr<::uwb::RegisteredCallbackToken> token = *tokenIt;
    // do whatever is needed with token
    // erase token from token container
    m_tokens.erase(tokenIt);
}
```

* It seems like we could store the callbacks themselves in `::uwb::RegisteredCallbackToken` since we know we're going to have one for each registration. So why not consolidate? The `RegisteredCallbackToken` then becomes more like a callback state tracking structure rather than just the token. Eg.

```
struct RegisteredCallbackToken;
struct DeviceRegisteredCallbackToken : public RegisteredCallbackToken {
    std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks> Callbacks;
};
struct SessionRegisteredCallbackToken : public RegisteredCallbackToken {
    std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks> Callbacks;
};

std::vector<std::shared_ptr<DeviceRegisteredCallbackToken>> m_deviceCallbacks;
std::unordered_map<uint32_t, std::shared_ptr<SessionRegisteredCallbackToken >> m_sessionEventCallbacks;
```

When you get a `std::weak_ptr<RegisteredCallbackToken>`, you'd need to use `std::dynamic_cast<T>` to determine if you've been passed a `DeviceRegisteredCallbackToken` or a `SessionRegisteredCallbackToken`.  

To avoid `dynamic_cast`, you could add a dedicated type field to the base:

```
enum class CallbackType { Device, Session };
struct RegisteredCallbackToken {
    CallbackType callbackType;
};
struct DeviceRegisteredCallbackToken : public RegisteredCallbackToken {
    DeviceRegisteredCallbackToken : RegisteredCallbackToken({ callbackType = CallbackType.Device}) {}
    std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks> Callbacks;
};
struct SessionRegisteredCallbackToken : public RegisteredCallbackToken {
    SessionRegisteredCallbackToken() : RegisteredCallbackToken({ callbackType = CallbackType.Session}) {}
    std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks> Callbacks;
};
```
Though, this is not as robust.

Last, you could avoid polymorphism all-together by using a `std::variant` in `RegisteredCallbackToken` instead:

```
struct RegisteredCallbackToken {
    std::variant<
        std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks>,
        std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks>> Callbacks;
};
```

This would require `std::visit` to determine the actual type, but that then allows compile-time determination using `constexpr`. I think this approach is pretty clean.

Can you consider these points for reducing complexity and consolidation and let me know whether you think any of them are worth doing?

Reply:
I didn't realize that the address of the token object could be used as a unique identifier, which is what the current design was built around. So, I'll make another PR that will address this fact, and I probably will use a lot of the suggestions here. Thanks!

Reply:
I just pushed some code that simplifies much of the code. I used the polymorphism approach, since I believe it makes the most sense given that we have separated session and device callbacks in the rest of the interface.

Reply:
Wow, this was really well done, I love it!